fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,147 @@
+    agent: Arc<IPCAgentProxy>,
+}
+
+impl<T> PollingParentSyncer<T> {
+    pub fn new(config: Config, parent_view_provider: Arc<T>, agent: Arc<IPCAgentProxy>) -> Self {
+        Self {
+            config,
+            started: Arc::new(AtomicBool::new(false)),
+            parent_view_provider,
+            agent,
+        }
+    }
+}
+
+impl<T: ParentFinalityProvider + Send + Sync + 'static> PollingParentSyncer<T> {
+    /// Start the proof of finality listener in the background
+    pub fn start(&mut self) -> anyhow::Result<()> {
+        match self
+            .started
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+        {
+            Ok(_) => {}
+            Err(_) => return Err(anyhow!("already started")),
+        }
+
+        let config = self.config.clone();
+        let provider = self.parent_view_provider.clone();
+        let agent = self.agent.clone();
+
+        tokio::spawn(async move {
+            if let Err(e) = sync_with_parent(config, agent, provider).await {
+                tracing::info!("sync with parent encountered error: {e}");
+            }
+        });
+
+        Ok(())
+    }
+}
+
+async fn sync_with_parent<T: ParentFinalityProvider + Send + Sync + 'static>(
+    config: Config,
+    agent_proxy: Arc<IPCAgentProxy>,
+    provider: Arc<T>,
+) -> anyhow::Result<()> {
+    let mut interval = tokio::time::interval(Duration::from_secs(config.polling_interval));
+
+    loop {
+        interval.tick().await;
+
+        // update block hash and validator set
+        let latest_height = agent_proxy
+            .get_chain_head_height()
+            .await
+            .context("cannot fetch parent chain head")?;
+
+        if latest_height < config.chain_head_delay {
+            tracing::debug!("latest height not more than the chain head delay");
+            continue;
+        }
+
+        let starting_height = if let Some(h) = provider.latest_height().await {
+            h + 1
+        } else {
+            // if latest_recorded_height is None, it means we have not started fetching any heights
+            // we just use the latest height minus a lower bound as the parent view
+            max(
+                1,
+                latest_height.saturating_sub(config.chain_head_lower_bound),
+            )
+        };
+
+        for h in starting_height..=latest_height {
+            let block_hash = agent_proxy
+                .get_block_hash(h)
+                .await
+                .context("cannot fetch block hash")?;
+            let validator_set = agent_proxy
+                .get_validator_set(h)
+                .await
+                .context("cannot fetch validator set")?;
+
+            match provider
+                .new_parent_view(Some((h, block_hash, validator_set)), vec![])
+                .await
+            {
+                Ok(_) => {}
+                Err(e) => match e {
+                    Error::ParentReorgDetected(_) => {
```

Comment:
I don't fully grasp how this error will come about. Say the `latest_height` is `N`, which means you will try to fetch from `N+1`. From the previous PR I [only see](https://github.com/consensus-shipyard/fendermint/pull/205/files#diff-8edd9923a9eb262950fd919a858ece4449d8a41abf401bd3c1e4ac1d63190454R88) that any error inserting into the height based cache is considered a reorg, which is if the height you are trying to insert isn't what's expected, ie. lower or higher than `N+1`. 

But since you are fetching from `N+1`, you won't see `N` or `N+2`. The only way to detect a reorg would be to make the cache more like a chain and check that the parent hash of the newly inserted head isn't what you had last, however this wouldn't even be possible because the cache doesn't store each height, only ones divisible by the "checkpointing" interval.

Reply:
Based on the way we are fetching, `N+2` should not have happened. Come to think of it, a reorg should already be detected if `starting_height > latest_height`.